### PR TITLE
UniqInjectiveFunctionsEliminationPass recursively removing chain of injective functions

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Eliminate injective function in argument of functions `uniq*` recursively. This used to work correctly but was broken in the new analyzer.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: Normal Builds
- [ ]  Allow: Special Builds
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by canhld94** [comment]
Should have performance improvement as well https://github.com/ClickHouse/ClickHouse/pull/65092#issuecomment-2162000302

```
0.502	0.111	-4.511x	-0.779	0.778	injective_functions_inside_uniq	1	SELECT uniq(hex(hex(number))) FROM numbers(50000000)
```

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit f8f585190ffc05da8011637cfa0e73a9317f84e2 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65140/f8f585190ffc05da8011637cfa0e73a9317f84e2/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by canhld94** [comment]
```
0.58	0.134	-4.325x	-0.769	0.768	injective_functions_inside_uniq	1	SELECT uniq(hex(hex(number))) FROM numbers(50000000)
0.193	0.134	-1.442x	-0.307	0.306	injective_functions_inside_uniq	0	SELECT uniq(negate(negate(negate(number)))) FROM numbers(50000000)
```

**Comment by Algunenano** [comment]
Please, don't mix a bugfix with a performance improvement and also style changes (`AggregateFunctionUniqCombined.cpp`). It makes it riskier to backport, so I'd rather ignore the PR completely.

**Comment by canhld94** [comment]
> Please, don't mix a bugfix with a performance improvement and also style changes (`AggregateFunctionUniqCombined.cpp`). It makes it riskier to backport, so I'd rather ignore the PR completely.

@Algunenano this is not a performance improvement, but rather a fix for performance degradation when enabling analyzer. It was tracked in https://github.com/ClickHouse/ClickHouse/issues/62245, but no one touch it yet.

Regarding the style change, it could be have been done automatically by clangd. I will change it back if necessary.

**Comment by Algunenano** [comment]
I would rather not backport a "performance degradation when enabling analyzer", I don't think it's worth the risk. But it's ok if you don't want to do it, I'll create a PR with just the fix for the crash (https://github.com/ClickHouse/ClickHouse/issues/65052) and backport it, and then we can iterate and add other improvements

**Comment by canhld94** [comment]
> I would rather not backport a "performance degradation when enabling analyzer", I don't think it's worth the risk. But it's ok if you don't want to do it, I'll create a PR with just the fix for the crash (#65052) and backport it, and then we can iterate and add other improvements

If we take a look at #65052, there are quite a number of PR for back porting performance degradation, so I don't really understand why we should treat this PR differently. Not to mention the implementation for fixing both issues are actually similar, so make it two PRs even make it more complicated IMO.

**Comment by Algunenano** [comment]
They are 2 different issues. As I said, you don't have to do it

**Comment by canhld94** [comment]
> They are 2 different issues. As I said, you don't have to do it

@Algunenano Thanks! I understand that you need to consider safety for back porting. What I'm trying to say here is I think this PR is relatively safe even for back porting (though the only 100% safe way is to just disable the optimization).

**Comment by Algunenano** [comment]
Yes, but I'd rather backport 2 different PRs doing 2 small and clear things, with clear tests, than a single one.

**Comment by canhld94** [comment]
Failing stateless, integration, and performance tests seem not relates:

```
03037_dynamic_merges_1_vertical
test_ttl_move/test.py::test_alter_with_merge_work[mt_work]
-- slowdown
SELECT count() FROM test.hits WHERE NOT ignore(countMatches(URL, '(?i)yandex|google')) SETTINGS max_threads=2
```

Performance test:

```
0.874 | 0.196 | -4.452x | -0.776 | 0.775 | injective_functions_inside_uniq | 1 | SELECT uniq(hex(hex(number))) FROM numbers(50000000)
0.238 | 0.196 | -1.214x | -0.177 | 0.176 | injective_functions_inside_uniq | 0 | SELECT uniq(negate(negate(negate(number)))) FROM numbers(50000000)
```

@novikd @Algunenano please help to review when you have time [THANKS]

**Comment by Algunenano** [comment]
I'll have a look. Thanks for the heads up

**Comment by novikd** [comment]
I'm not sure if it's a performance improvement, because it's fixes the optimisation we forgot in 24.3. IMHO it should be labeled as a bug fix and back ported to versions with enabled analyzer.

File: tests/queries/0_stateless/02493_analyzer_uniq_injective_functions_elimination.sql [hunk 1]
```diff
@@ -1,5 +1,9 @@
-SET allow_experimental_analyzer = 1;
+SET allow_experimental_analyzer = 1, optimize_injective_functions_inside_uniq = 1;
 
 EXPLAIN QUERY TREE SELECT uniqCombined(tuple('')) FROM numbers(1);
 
 SELECT uniqCombined(tuple('')) FROM numbers(1);
+
+EXPLAIN QUERY TREE SELECT uniqCombined(tuple(materialize(tuple(number)))) FROM numbers(10);
+
+SELECT uniqCombined(tuple(materialize(tuple(number)))) FROM numbers(10);
```
Comment (Algunenano) [inline tests/queries/0_stateless/02493_analyzer_uniq_injective_functions_elimination.sql:9]: The lack of tests for this feature is extremely sad to me (not your fault, it was bad already). 0 surprises it did not work properly, and 0 surprises if it breaks again in the future.

Do you mind adding tests that check `non-injective` functions and multi-argument functions too, please? It'd be a really great addition.

Comment (Algunenano) [reply tests/queries/0_stateless/02493_analyzer_uniq_injective_functions_elimination.sql]: Thank you @canhld94

**Review by Algunenano — APPROVED** [review]

**Review by novikd — APPROVED** [review]

**Pull request merged**